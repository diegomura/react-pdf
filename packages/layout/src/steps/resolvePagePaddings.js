import * as R from 'ramda';

import matchPercent from '../utils/matchPercent';

/*
 * Translates page percentage horizontal paddings in fixed ones
 *
 * @param {Object} page container
 * @param {String} padding value
 * @returns {Object} translated padding value
 */
const resolvePageHorizontalPadding = container => value => {
  const match = matchPercent(value);
  return match ? match.percent * container.width : value;
};

/**
 * Translates page percentage vertical paddings in fixed ones
 *
 * @param {Object} page container
 * @param {String} padding value
 * @returns {Object} translated padding value
 */
const resolvePageVerticalPadding = container => value => {
  const match = matchPercent(value);
  return match ? match.percent * container.height : value;
};

/**
 * Translates page percentage paddings in fixed ones
 *
 * @param {Object} page
 * @returns {Object} page with fixed paddings
 */
const resolvePagePaddings = page => {
  const container = page.style;

  return R.evolve({
    style: R.evolve({
      paddingTop: resolvePageVerticalPadding(container),
      paddingLeft: resolvePageHorizontalPadding(container),
      paddingRight: resolvePageHorizontalPadding(container),
      paddingBottom: resolvePageVerticalPadding(container),
    }),
  })(page);
};

/**
 * Translates all pages percentage paddings in fixed ones
 * This has to be computed from pages calculated size and not by Yoga
 * because at this point we didn't performed pagination yet.
 *
 * @param {Object} document root
 * @returns {Object} document root with translated page paddings
 */
export default R.evolve({
  children: R.map(resolvePagePaddings),
});
