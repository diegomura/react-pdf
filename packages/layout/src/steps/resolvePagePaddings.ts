import { evolve, matchPercent } from '@react-pdf/fns';
import { SafeStyle } from '@react-pdf/stylesheet';

import { SafeDocumentNode, SafePageNode } from '../types';

/**
 * Translates page percentage horizontal paddings in fixed ones
 *
 * @param container - Page container
 * @returns Resolve page horizontal padding
 */
const resolvePageHorizontalPadding =
  (container: SafeStyle) => (value: number) => {
    const match = matchPercent(value);
    const width = container.width as number;
    return match ? match.percent * width : value;
  };

/**
 * Translates page percentage vertical paddings in fixed ones
 *
 * @param container - Page container
 * @returns Resolve page vertical padding
 */
const resolvePageVerticalPadding =
  (container: SafeStyle) => (value: number) => {
    const match = matchPercent(value);
    const height = container.height as number;
    return match ? match.percent * height : value;
  };

/**
 * Translates page percentage paddings in fixed ones
 *
 * @param page
 * @returns Page with fixed paddings
 */
const resolvePagePaddings = (page: SafePageNode): SafePageNode => {
  const container = page.style;

  const style = evolve(
    {
      paddingTop: resolvePageVerticalPadding(container),
      paddingLeft: resolvePageHorizontalPadding(container),
      paddingRight: resolvePageHorizontalPadding(container),
      paddingBottom: resolvePageVerticalPadding(container),
    },
    page.style,
  );

  return Object.assign({}, page, { style });
};

/**
 * Translates all pages percentage paddings in fixed ones
 * This has to be computed from pages calculated size and not by Yoga
 * because at this point we didn't performed pagination yet.
 *
 * @param root - Document root
 * @returns Document root with translated page paddings
 */
const resolvePagesPaddings = (root: SafeDocumentNode) => {
  if (!root.children) return root;

  const children = root.children.map(resolvePagePaddings);

  return Object.assign({}, root, { children });
};

export default resolvePagesPaddings;
