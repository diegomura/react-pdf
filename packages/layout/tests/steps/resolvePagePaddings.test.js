import resolvePagePaddings from '../../src/steps/resolvePagePaddings';

describe('layout resolvePagePaddings', () => {
  test('Should keep other styles untouched', () => {
    const root = {
      type: 'DOCUMENT',
      children: [{ type: 'PAGE', style: { color: 'red' } }],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingTop as it is', () => {
    const root = {
      type: 'DOCUMENT',
      children: [{ type: 'PAGE', style: { paddingTop: 50 } }],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingRight as it is', () => {
    const root = {
      type: 'DOCUMENT',
      children: [{ type: 'PAGE', style: { paddingRight: 50 } }],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingBottom as it is', () => {
    const root = {
      type: 'DOCUMENT',
      children: [{ type: 'PAGE', style: { paddingBottom: 50 } }],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingLeft as it is', () => {
    const root = {
      type: 'DOCUMENT',
      children: [{ type: 'PAGE', style: { paddingLeft: 50 } }],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingTop', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: { paddingTop: '10%', width: 100, height: 200 },
        },
      ],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingRight', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: { paddingRight: '10%', width: 100, height: 200 },
        },
      ],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingBottom', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: { paddingBottom: '10%', width: 100, height: 200 },
        },
      ],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingLeft', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: { paddingLeft: '10%', width: 100, height: 200 },
        },
      ],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });

  test('Should resolve several pages', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: { paddingTop: 10, width: 100, height: 200 },
        },
        {
          type: 'PAGE',
          style: { paddingBottom: '10%', width: 100, height: 200 },
        },
        {
          type: 'PAGE',
          style: {
            paddingRight: 10,
            paddingLeft: '10%',
            width: 100,
            height: 200,
          },
        },
      ],
    };
    const result = resolvePagePaddings(root);

    expect(result).toMatchSnapshot();
  });
});
