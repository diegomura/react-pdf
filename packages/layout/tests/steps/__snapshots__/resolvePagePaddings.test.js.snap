// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout resolvePagePaddings Should keep other styles untouched 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "color": "red",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should leave numeric paddingBottom as it is 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "paddingBottom": 50,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should leave numeric paddingLeft as it is 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "paddingLeft": 50,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should leave numeric paddingRight as it is 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "paddingRight": 50,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should leave numeric paddingTop as it is 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "paddingTop": 50,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should resolve percent paddingBottom 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "height": 200,
        "paddingBottom": 20,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should resolve percent paddingLeft 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "height": 200,
        "paddingLeft": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should resolve percent paddingRight 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "height": 200,
        "paddingRight": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should resolve percent paddingTop 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "height": 200,
        "paddingTop": 20,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings Should resolve several pages 1`] = `
Object {
  "children": Array [
    Object {
      "style": Object {
        "height": 200,
        "paddingTop": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
    Object {
      "style": Object {
        "height": 200,
        "paddingBottom": 20,
        "width": 100,
      },
      "type": "PAGE",
    },
    Object {
      "style": Object {
        "height": 200,
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;
